src/visitors/sympy_replace_solutions_visitor.hpp

```diff
@@ -0,0 +1,465 @@
+         * what equations need to be updated when I change a particular variable. To do that we
+         * build a a graph of dependencies var -> vars and in the mean time we reduce it to the root
+         * variables. This is ensured by the fact that the tmp variables are sorted so that the next
+         * tmp variable may depend on the previous one. Since it is a relation of equivalence (if an
+         * equation depends on a variable, it needs to be updated if the variable changes), we build
+         * the two maps at the same time.
+         *
+         * An example:
+         *
+         *  - \f tmp0 = x + a \f
+         *  - \f tmp1 = tmp0 + b \f
+         *  - \f tmp2 = y \f
+         *
+         * dependency_map_ should be (the order of the equation is unimportant since we are building
+         * a map):
+         *
+         * - tmp0 : x, a
+         * - tmp1 : x, a, b
+         * - tmp2 : y
+         *
+         * and the var2statement_ map should be (the order of the following equations is unimportant
+         * since we are building a map. The number represents the index of the original equations):
+         *
+         * - x : 0, 1
+         * - y : 2
+         * - a : 0, 1
+         * - b : 1
+         */
+        void build_maps();
+
+        /// Check if one of the statements assigns this variable (i.e. \f x' = f(x, y, x) \f) and is
+        /// still tagged
+        inline bool is_var_assigned_here(const std::string& var) const {
+            const auto it = var2statement_.find(var);
+            return it != var2statement_.end() && tags_.find(it->second) != tags_.end();
+        }
+
+        /**
+         * \brief Look for \p var in \ref var2statement_ and emplace back that statement in \p
+         * new_statements
+         *
+         * If there is no \p var key in \ref var2statement_, return false
+         */
+        bool try_emplace_back_tagged_statement(ast::StatementVector& new_statements,
+                                               const std::string& var);
+
+
+        /// Emplace back the next \p n_next_statements solutions in \ref statements that is marked
+        /// for updating in \ref tags_
+        size_t emplace_back_next_tagged_statements(ast::StatementVector& new_statements,
+                                                   const size_t n_next_statements);
+
+        /// Emplace back all the statements that are marked for updating in \ref tags_
+        size_t emplace_back_all_tagged_statements(ast::StatementVector& new_statements);
+
+        /**
+         * \brief Tag all the statements that depend on \p var for updating
+         *
+         * This is necessary when an assignment has invalidated this variable
+         */
+        size_t tag_dependant_statements(const std::string& var);
+
+        /// Mark that all the statements need updating (probably unused)
+        void tag_all_statements();
+
+        /**
+         * \brief var -> (depends on) vars
```

Comment:
nitpick : this `brief` explanation could be better :)

Reply:
done